Removing toy benchmark, adding comparison with native system call times

JustinCappos opened the pull request:
## Description

Fixes # (issue)

<!-- Please include a summary of the changes and the related issue. --> 
<!-- Please also include relevant motivation and context. Why is this change required? What problem does it solve? -->
<!-- List any dependencies that are required for this change. -->

### Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->

- Test A - `lind_project/tests/test_cases/test_a.c`
- Test B - `lind_project/tests/test_cases/test_b.c`

## Checklist:

<!-- Add details about the checklist whenever needed -->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

JustinCappos committed:
Removing toy benchmark, adding comparison with Native

rennergade commented on benches/rustposix-criterion.rs:
```diff
@@ -46,14 +36,20 @@ fn get_dummy_cage() -> Cage {
-    c.bench_function("fib 20", |b| b.iter(|| fibonacci(black_box(20))));
+    // I'm not setting this up properly because I think I don't need that much
+    // state.  I just want to call system calls.
     let a = get_dummy_cage();
```
In our tests we do lindrustinit/finalize to setup and shutdown, this will create an initial cage. That may work here as well unless you're trying to test for a specific state. Not partial to either, just pointing out the difference.

rennergade reviewed (commented):
I like this a lot. Seems self-explanatory and very useful. I didn't see any necessary changes and just left a minor suggestion.

Yaxuan-w reviewed (approved):
Code is easy to understand and test makes sense to me

yzhang71 reviewed (approved):
LGTM!

JustinCappos committed:
adding more tests

JustinCappos commented:
> I left a minor suggestion.

I added a fix to address this in my latest commit.  I also added a few more tests.  

Could someone take a look and if it's okay, go ahead and merge?  Please be warned, that I really don't get some aspects of Rust or Lind file system initialization + finalization so, please double check for bugs...

Yaxuan-w commented on benches/rustposix-criterion.rs:
```diff
@@ -1,68 +1,157 @@
+                let _ = libc::write(fd,CString::new("Well, hello there!!!").unwrap().as_ptr() as *const c_void,20);
+                let mut read_buffer = sizecbuf(20);
+                libc::lseek(fd,0,SEEK_SET);
+                cage.read_syscall(fd,read_buffer.as_mut_ptr(), 20);
```
If we want to test Native OS, would using `libc::read(fd, read_buffer.as_mut_ptr(), 20)` makes more sense here?

JustinCappos replied:
```suggestion
                libc::read(fd,read_buffer.as_mut_ptr() as *mut c_void, 20);
```

JustinCappos replied:
Thanks, fixed!

JustinCappos committed:
Update benches/rustposix-criterion.rs

Fix to use native call instead of Lind call

rennergade commented on benches/rustposix-criterion.rs:
```diff
@@ -1,68 +1,157 @@
+ * results checking / assertations to avoid adding bias to the results.  */
+
+
+// I hate allowing this, but this is apparently a known issue for a lot of 
```
There are some helper functions in tests/mod.rs that I believe would help you do these more safely.

JustinCappos replied:
I think I addressed this.  I use those functions now.

rennergade commented on benches/rustposix-criterion.rs:
```diff
@@ -1,68 +1,157 @@
+    group.bench_function("Native OS kernel read+write", |b| b.iter(|| 
+        {
+            unsafe{
+                let _ = libc::write(fd,CString::new("Well, hello there!!!").unwrap().as_ptr() as *const c_void,20);
```
I think in general declaring the strings outside of the unsafe blocks/function signature would also help with safety/clarity

JustinCappos replied:
I tried what you recommended and it doesn't remove the need to have: #![allow(temporary_cstring_as_ptr)]

To me, it doesn't matter one way or the other, but I'm not enough in touch with what the right style is to really judge.

`fd = libc::open(CString::new("/tmp/foo").unwrap().as_ptr(),O_CREAT | O_TRUNC | O_WRONLY,S_IRWXA);`

versus

```
    let filename = CString::new("/tmp/foo").unwrap().as_ptr();
    unsafe {
        fd = libc::open(filename,O_CREAT | O_TRUNC | O_WRONLY,S_IRWXA);
    }
```

seems pretty similar to me.

JustinCappos replied:
Okay, I've updated the PR to move to the str2cbuf(...) function instead everywhere.  (Loosely related.)

JustinCappos committed:
Benchmark cleanup

JustinCappos committed:
Switching to our str2cbuf

JustinCappos committed:
Numbering benchmarks because this helps us run them selectively

JustinCappos commented:
@rennergade This should be ready now.  Please squash and merge if you think it's ready.

rennergade reviewed (approved):
Awesome this looks great.
